Add QubitSparsePauli and QubitSparsePauliList classes to quantum_info

DanPuzzuoli opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Edit: This is the first PR in a series of three introducing `QubitSparsePauli`, `QubitSparsePauliList`, and `PauliLindbladMap`. This PR is primarily about introducing the data structure for `QubitSparsePauli` and `QubitSparsePauliList`.

This PR adds `QubitSparsePauli` and `QubitSparsePauliList` representing, respectively, a Pauli operator and a list of Pauli operators in qubit-sparse format. These classes are essentially copies of `SparseTerm` and `SparseObservable`, with the changes:
- The alphabet is restricted to the symbols `"IXYZ"`
- They do not have coefficients (they do not represent sums or terms in sums, just qubit-sparse Paulis and lists of qubit-sparse Paulis)

The choice to develop these classes was made during the first attempt at implementing `PauliLindbladMap`; see the closed and incomplete PR #14260 . `PauliLindbladMap` required storing a qubit-sparse list of paulis, but we realized that we want the user to be able to construct qubit-sparse Paulis and qubit-sparse Pauli lists directly to be used as method arguments (e.g. a `pauli_fidelity` method) or to be able to return a qubit-sparsely specified list of Paulis (e.g. a `sample` method which could return a list of qubit-sparse Paulis when drawing multiple samples). As such it made sense to make `QubitSparsePauli` and `QubitSparsePauliList` as standalone objects. I've put these in a `pauli_lindblad_map` folder and rust module in anticipation of also adding a `PauliLindbladMap` class there.

In a subsequent PR, I will remake `PauliLindbladMap`, and it will make use of `QubitSparsePauliList` to internally store the list of operators.

### Details and comments

As mentioned these classes are nearly direct copies of `SparseTerm` and `SparseObservable`, but with a lot of minor docs changes, changes to names of methods (e.g. `SparseObservable.zero` was renamed to `QubitSparsePauliList.empty`), and in some cases removal of functionality that doesn't make sense for a "list" (e.g. `SparseObservable.identity` was deleted). Some more detailed (but non-exhaustive) notes:

- SparseObservable renamed QubitSparsePauliList
   - Removed “identity”, zero → empty
- SparseTerm renamed QubitSparsePauli
   - Modified constructors to be similar in format to QubitSparsePauliList. I.e. previously PyQubitSparsePauli.py_new was like PyQubitSparsePauliList.from_raw_parts. I’ve added from_pauli, from_sparse_label, from_label, changed py_new → from_raw_parts, and made the default constructor delegate to the others (similar to PyQubitSparsePauliList)
   - Note that this class lacks some capabilities of QubitSparsePauliList. E.g. the attributes are not writable, and they return raw arrays (as opposed to views)
- Removed all non-Pauli characters
   - removed functions like “pauli_bases” that only make sense in the context with non-Pauli characters

At the moment I couldn't get the docs to build locally due to some issues with `doxygen`.

DanPuzzuoli committed:
attempting to add initial structure

DanPuzzuoli committed:
adding module creation function

DanPuzzuoli committed:
partial progress with copying

DanPuzzuoli committed:
partial progress 2

DanPuzzuoli committed:
partial progress 3

DanPuzzuoli committed:
copying over

DanPuzzuoli committed:
successful compilation and import

DanPuzzuoli committed:
minor changes

DanPuzzuoli committed:
renaming SparseTerm to QubitSparsePauli

DanPuzzuoli committed:
starting to transfer tests

DanPuzzuoli committed:
moving more tests

DanPuzzuoli committed:
updating repr of QubitSparsePauliList and adding tests

DanPuzzuoli committed:
testing from_pauli

DanPuzzuoli committed:
more test progress

DanPuzzuoli committed:
moved relevant QubitSparsePauliList tests

DanPuzzuoli committed:
adding from_label and from_pauli constructors to QubitSparsePauli

DanPuzzuoli committed:
adding tests from from_pauli and from_label

DanPuzzuoli committed:
added from_sparse_label method

DanPuzzuoli committed:
fixing QubitSparsePauli constructors

DanPuzzuoli committed:
working through tests for QubitSparsePuli

DanPuzzuoli committed:
saving

DanPuzzuoli committed:
done adding tests (I think)

DanPuzzuoli committed:
formatting and linting

DanPuzzuoli committed:
QubitSparsePauli main doc

DanPuzzuoli committed:
done initial pass on QubitSparsePauliList docs

DanPuzzuoli committed:
initial pass on docs for QubitSparsePauli

DanPuzzuoli committed:
last edits

SamFerracin commented on test/python/quantum_info/test_qubit_sparse_pauli.py:
```diff
@@ -0,0 +1,1333 @@
+        with self.assertRaisesRegex(ValueError, "explicitly given 'num_qubits'"):
+            QubitSparsePauli(data, num_qubits=data.num_qubits + 1)
+
+        with_phase = Pauli("-jIYYXY")
```
Oh, you went for ignoring the phases then. Let's make sure that everyone sees this before we merge, and speak now or forever hold your peace

jakelishman replied:
This feels fine to me if explicitly documented, which it is.

SamFerracin commented on test/python/quantum_info/test_qubit_sparse_pauli.py:
```diff
@@ -0,0 +1,1333 @@
+        )
+
+    def test_from_raw_parts(self):
+        # Happiest path: exactly typed inputs.
```
What do you mean here?

DanPuzzuoli replied:
This was copied over from the `SparseObservable` tests. This test case tests various input types that are flagged with comments, and I think this comment is just flagging that the first test block is for when the inputs are supplied in the ideal type.

SamFerracin commented on crates/accelerate/src/pauli_lindblad_map/mod.rs:
```diff
@@ -0,0 +1,22 @@
+pub mod qubit_sparse_pauli;
+
+pub fn pauli_lindblad_map(m: &Bound<PyModule>) -> PyResult<()> {
+    m.add_class::<PyQubitSparsePauliList>()?;
```
Check if qiskit tends to use alphabetical order. If so, switch these two

DanPuzzuoli replied:
In general I believe it does, so I'll switch them.

jakelishman replied:
The order you add stuff here doesn't really matter - if we're usually alphabetical, it's probably just individual people all separately choosing that for aesthetic reasons.

SamFerracin commented on crates/accelerate/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,2466 @@
+/// `TypeError`-related ones are statically forbidden (within Rust) by the language, and conversion
+/// failures on entry to Rust from Python space will automatically raise `TypeError`.
+#[derive(Error, Debug)]
+pub enum CoherenceError {
```
I assume you use them all?

DanPuzzuoli replied:
I think the only one not used in this PR is the first one, but it will be used by `PauliLindbladMap` in the subsequent PR.

SamFerracin commented on crates/accelerate/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,2466 @@
+///   ==================  ===========  =============================================================
+///   Attribute           Length       Description
+///   ==================  ===========  =============================================================
+///   :attr:`bit_terms`   :math:`s`    Each of the non-identity single-qubit Pauli operators.  These
```
I would prefer something less computer-sciency and more quantum, like "pauli_terms." And if so, I would rename also the rust class. What do you think?

DanPuzzuoli replied:
I would vote to keep it as is for interface consistency with `SparseObservable`.

DanPuzzuoli replied:
Update: The C tests are currently failing due to a naming collision between the `BitTerm` here and the `BitTerm` in `sparse_observable/mod.rs`. Even if we can fix this technically and keep the name, this adds a point in favour of changing the name: the naming collision could be confusing.

DanPuzzuoli replied:
Just renamed `BitTerm` to `Pauli` based on suggestion by @jakelishman

SamFerracin commented on crates/accelerate/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,2466 @@
+///   ============================  ================================================================
+///   Method                        Summary
+///   ============================  ================================================================
+///   :meth:`from_label`            Convert a dense string label into a :class:`.QubitSparsePauli`.
```
```suggestion
///   :meth:`from_label`            Convert a dense string label into a :class:`~.QubitSparsePauli`.
```
You may need this for the hyperliink to work, not sure though

SamFerracin commented on crates/accelerate/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,2466 @@
+///   ============================  ================================================================
+///   :meth:`from_label`            Convert a dense string label into a :class:`.QubitSparsePauli`.
+///
+///   :meth:`from_sparse_label`     A tuple of a sparse string label and the qubits they apply to.
```
```suggestion
///   :meth:`from_sparse_label`     Build a :class:`.QubitSparsePauli` from a tuple of a sparse string label and the qubits they apply to.
```

DanPuzzuoli committed:
Update crates/accelerate/src/pauli_lindblad_map/qubit_sparse_pauli.rs

Co-authored-by: Samuele Ferracin <sam.ferracin@ibm.com>

DanPuzzuoli committed:
Update crates/accelerate/src/pauli_lindblad_map/qubit_sparse_pauli.rs

Co-authored-by: Samuele Ferracin <sam.ferracin@ibm.com>

DanPuzzuoli committed:
changing the order of the class additions in mod.rs

DanPuzzuoli committed:
attempting to fix table

DanPuzzuoli committed:
fixing last warning in docs build

DanPuzzuoli committed:
renaming BitTerm -> Pauli

DanPuzzuoli committed:
fixing formatting

DanPuzzuoli committed:
fixing table indentation

DanPuzzuoli committed:
fixing error message

DanPuzzuoli committed:
making error checking test more specific

DanPuzzuoli committed:
cleaning up some docs

DanPuzzuoli committed:
fixing test

jakelishman reviewed (commented):
Overall, this looks like it'd be generally fine to me.  I don't love the amount of code duplication from `SparseObservable`, but I also don't have any reasonable alternative.

I appreciate the effort you went to to strip the concept of "arithmetic" from the object, since it's list-like and not sum-like, both in any remnant method names and the documentation.

One question about the structure: since you're making a "collection" not a "sum": do you think that a set-like interface (with automatic de-duplication) would be most appropriate for you, or should duplicates definitely be allowed?  `SparseObservable` uses the list-like form because it was optimising to allow direct list-like access into the fields even from Python space, in part to allow fast arithmetic operations.  I suspect that that might be less of a concern for you, and you'll care more about union/intersection operations?

As a follow-on from that: do you _need_ to expose the backing data structures (`boundaries` and the flat lists `indices` and `paulis`) to Python space?  Just asking because if there's no need and you can provide higher-level interfaces only, I'd keep the internals private - that a) simplifies the code a whole bunch, and b) gives us way more flexibility in the future.

I haven't reviewed #14328 in any real detail, but some considerations from having scanned it:
- I see that you make all the internal fields of `QubitSparsePauliList` `pub`.  I'd be _really_ cautious about that, because it allows users of the struct to rely on the internals, and to make arbitrary changes to them, which can destroy data coherence really easily.  I suspect that many of the uses of that can be wrapped up into safe interfaces (like iterating over terms, etc), and hopefully all the remaining cases can just have extra Rust-space methods as well.  If that's not possible, it might indicate an API-boundary problem, so it's worth consideration now.
- Given how `PauliLindbladMap` associates a rate with a `QubitSparsePauliList`, that does feel like a set-like backing might be appropriate?

The Sphinx docs errors look like they were mostly caused by the naming clash with `qiskit.quantum_info.Pauli`.  The solution will be to replace anywhere (other than this file) in Qiskit that writes `` :class:`.Pauli` `` with `` :class:`~.quantum_info.Pauli` `` which should do it.  It's slightly annoying busy-work, but it's something we have to do with other classes already.

DanPuzzuoli commented:
> One question about the structure: since you're making a "collection" not a "sum": do you think that a set-like interface (with automatic de-duplication) would be most appropriate for you, or should duplicates definitely be allowed? `SparseObservable` uses the list-like form because it was optimising to allow direct list-like access into the fields even from Python space, in part to allow fast arithmetic operations. I suspect that that might be less of a concern for you, and you'll care more about union/intersection operations?

So we need `QubitSparsePauliList` to:
- Maintain ordering in relation to `rates` for `PauliLindbladMap`. My understanding is that any set construction wouldn't come with a guaranteed ordering? (See the comment on the rate <-> Pauli association below.)
- It's also going to be used within the return type of `PauliLindbladMap.sample`, which will return `n` samples from the pseudo-probability representation of the Pauli Lindblad map (each "sample" is of type `(bool, QubitSparsePauli)`). In this case duplication is possible, and again here the operators all need to be associated with another value (a boolean) so it seems like some maintenance of ordering is necessary.

> As a follow-on from that: do you _need_ to expose the backing data structures (`boundaries` and the flat lists `indices` and `paulis`) to Python space? Just asking because if there's no need and you can provide higher-level interfaces only, I'd keep the internals private - that a) simplifies the code a whole bunch, and b) gives us way more flexibility in the future.

I think probably yes... otherwise I'm not sure how a user would even know what data the object contains? I guess they could iterate through it and look at the underlying `QubitSparsePauli`s? Tbh it isn't immediately clear if this is necessary, but it seems odd to not be able to look inside. 

> I haven't reviewed #14328 in any real detail, but some considerations from having scanned it:
> 
>     * I see that you make all the internal fields of `QubitSparsePauliList` `pub`.  I'd be _really_ cautious about that, because it allows users of the struct to rely on the internals, and to make arbitrary changes to them, which can destroy data coherence really easily.  I suspect that many of the uses of that can be wrapped up into safe interfaces (like iterating over terms, etc), and hopefully all the remaining cases can just have extra Rust-space methods as well.  If that's not possible, it might indicate an API-boundary problem, so it's worth consideration now.

I can think about this when progressing to that PR - I'll need to remove the `pub` and see what breaks. Tbh I think I just made these `pub` to access them easily (as a rust newb) so it probably won't be hard to remove.

>     * Given how `PauliLindbladMap` associates a rate with a `QubitSparsePauliList`, that does feel like a set-like backing might be appropriate?
> 

A rate is not associated with a `QubitSparsePauliList` - `rates` is a list of the same length as `QubitSparsePauliList`, and `rates[i]` is associated with `qubit_sparse_pauli_list[i]` (the ordering matters here). There may be some confusion from my original proposal, which did have a set associated with each rate. We ended up deciding to drop this. In a subsequent PR I will add `simplify()` (like for `SparseObservable`) for collecting terms with the same operator.

> The Sphinx docs errors look like they were mostly caused by the naming clash with `qiskit.quantum_info.Pauli`. The solution will be to replace anywhere (other than this file) in Qiskit that writes `` :class:`.Pauli` `` with `` :class:`~.quantum_info.Pauli` `` which should do it. It's slightly annoying busy-work, but it's something we have to do with other classes already.

Okay I'll fix this, thanks.

DanPuzzuoli committed:
changing all docs references to standard quantum_info to :class:

DanPuzzuoli committed:
updating all pauli references

DanPuzzuoli committed:
removing from_raw_parts and raw array data properties from QubitSparsePauliList

DanPuzzuoli committed:
deleting documentation for QubitSparsePauliList that details the internal array storage

DanPuzzuoli committed:
removing the Pauli attribute from QubitSparsePauliList

DanPuzzuoli committed:
fixing tests

DanPuzzuoli committed:
re-adding Pauli alphabet to QubitSparsePauli

DanPuzzuoli committed:
formatting

DanPuzzuoli commented:
In the latest round of commits I've removed the `paulis`, `indices`, and `boundaries` attributes of `QubitSparsePauliList`, along with the `QubitSparsePauliList.from_raw_parts` constructor (which no longer makes sense as the user won't know what the "raw parts" are), as well as the `Pauli` attribute from `QubitSparsePauliList`. To coincide with this, I've changed the class documentation to no longer describe the internal array storage methods.  As it is no longer needed, I've removed all of the `ArrayView` machinery.

I've tried the following:

> The Sphinx docs errors look like they were mostly caused by the naming clash with `qiskit.quantum_info.Pauli`. The solution will be to replace anywhere (other than this file) in Qiskit that writes `` :class:`.Pauli` `` with `` :class:`~.quantum_info.Pauli` `` which should do it. It's slightly annoying busy-work, but it's something we have to do with other classes already.

but am still locally getting some docs build errors that are unclear, e.g.: 

```
/Users/danielpuzzuoli/Documents/software_projects/qiskit/.tox/docs/lib/python3.13/site-packages/qiskit/result/sampled_expval.py:
docstring of qiskit.result.sampled_expval.sampled_expectation_value:1: 
WARNING: more than one target found for cross-reference 'Pauli': qiskit.quantum_info.Pauli,
 qiskit.quantum_info.operators.symplectic.pauli.Pauli, 
qiskit.quantum_info.QubitSparsePauli.Pauli
```

Interestingly this file does no documentation referencing to `quantum_info.Pauli`, though it does include `Pauli` in a type hint.

DanPuzzuoli committed:
minor docs change

DanPuzzuoli committed:
removing IterMut

DanPuzzuoli committed:
removing unnecessary methods return mutable references

DanPuzzuoli committed:
fixing method name

DanPuzzuoli committed:
linting

jakelishman committed:
Merge remote-tracking branch 'ibm/main' into qubit-sparse-pauli

The principle conflict is caused by quantum-info-related modules moving
into a separate `qiskit_quantum_info` crate, which this merge resolution
does.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,1901 @@
+            Self::Y => "Y",
+            Self::Z => "Z",
+        }
+    }
```
These are always going to be identical for you with `Pauli` (unlike `BitTerm`, which has `+` and `PLUS`, for example).  You only need one of the two.

DanPuzzuoli replied:
Dropped `py_name`.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,1901 @@
+
+    /// Get an iterator over the individual elements of the list.
+    pub fn iter(&'_ self) -> impl ExactSizeIterator<Item = QubitSparsePauliView<'_>> + '_ {
+        std::ops::Range {
+            start: 0,
+            end: self.boundaries.len() - 1,
+        }
```
This is fine, but for the future, it has a shorthand spelling: `(0..self.boundaries.len() - 1)`

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,1901 @@
+
+    fn to_label(&self) -> PyResult<String> {
+        Ok(self.inner.view().to_sparse_str())
+    }
```
Do you very much need this method?  I'm concerned that its definition of "label" is not the same as `from_label`, which is likely confusing (and it's not documented).  I'd suggest removal - the `to_sparse_str` internal in `SparseObservable` was only really meant to be used by the user-readable `__repr__`.

DanPuzzuoli replied:
Nope, removed.

jakelishman commented on crates/quantum_info/src/pauli_lindblad_map/qubit_sparse_pauli.rs:
```diff
@@ -0,0 +1,1901 @@
+            .map(|bit| bit.py_label())
+            .collect();
+        PyString::new(py, string.as_str())
+    }
```
I'm not sure how this method is intended to be used, because it's only going to return the non-identity terms.

If you need to include it, I'd suggest including a documentation example that illustrates how you pair it with `indices` to make an input for `from_sparse_list`.

jakelishman replied:
Also, if you do keep it, it needs tests.

DanPuzzuoli replied:
Dropped this too.

jakelishman committed:
Correct extraneous documentation

jakelishman commented on test/python/quantum_info/test_qubit_sparse_pauli.py:
```diff
@@ -0,0 +1,989 @@
+        # The label is interpreted like a bitstring, with the right-most item associated with qubit
+        # 0, and increasing as we move to the left (like `Pauli`, and other bitstring conventions).
+        self.assertEqual(
+            # Ruler for counting terms:  dcba9876543210
+            QubitSparsePauli.from_label("IXXIIZZIYYIXYZ"),
```
Glad to see my "ruler" survived 😂

jakelishman commented on test/python/quantum_info/test_qubit_sparse_pauli.py:
```diff
@@ -0,0 +1,989 @@
+        # The purpose of this is just to test that the `repr` doesn't crash, rather than asserting
+        # that it has any particular form.
+        self.assertIsInstance(repr(data), str)
+        self.assertIn("QubitSparsePauli", repr(data))
```
I don't think this comment is accurate (it's missing the empty Pauli), but also both "terms" and "non-identities per term" would only apply to `QubitSparsePauliList`, not `QubitSparsePauli`.

DanPuzzuoli replied:
Nice catch - I dropped the comment, and added an additional test case
```
QubitSparsePauli.from_label("III")
```

jakelishman commented on test/python/quantum_info/test_qubit_sparse_pauli.py:
```diff
@@ -0,0 +1,989 @@
+        QubitSparsePauliList.from_label("X"),
+        QubitSparsePauliList.from_list(["YIXZII"]),
+        QubitSparsePauliList.from_list(["YIXZII", "ZZYYXX"]),
+    )
```
Not sure the comment is accurate here either, but then again, I wouldn't 100% swear I was correct in my original either.

DanPuzzuoli replied:
Likewise dropped the comment here, and added the test case:
```
QubitSparsePauliList.from_list(["IIIIII", "ZZYYXX"])
```
(adding a case with an identity operator).

jakelishman reviewed (commented):
This was a slightly surreal experience, since so much of this (especially the test suite!) was originally stuff I wrote haha.

I had a couple of minor questions, but nothing major about the interfaces from this point.  I fixed the merge conflicts (#14342 moved where the file should be - note it's in `crates/quantum_info` now), and a minor documentation hold-over from `SparseObservable` that I spotted.

Happy to merge based on answers to the last couple of questions.

DanPuzzuoli committed:
removing Pauli.py_name and changing all calls to it to py_label

DanPuzzuoli committed:
dropping other unused methods

DanPuzzuoli committed:
fixing test comments/cases

DanPuzzuoli committed:
formatting

DanPuzzuoli commented:
Thanks for the thorough review Jake! I've dropped the methods you mentioned, removed the comments, and rerun the formatting.

jakelishman reviewed (approved):
Ace, thanks Dan - this looks like a solid base to me.

We need to remember to include a release note before the actual release, but we can do that in the last PR of the series. (*edit*: now I see you've got the base of one in the later PRs - nice.)
